src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1585,4 +1604,277 @@ public Text getUsage(CommandSource src) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            if (this.returnURI) {
+                return uri;
+            } else {
+                return url;
+            }
+        }
+    }
+
+    /**
+     * Expect an argument to return an IP address, in the form of an
+     * {@link InetAddress}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ip(Text key) {
+        return new IpElement(key, false);
+    }
+
+    /**
+     * Expect an argument to return an IP address, in the form of an
+     * {@link InetAddress}, or if nothing matches and the source is a
+     * {@link RemoteSource}, return the source's address.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ipOrSource(Text key) {
+        return new IpElement(key, true);
+    }
+
+    private static class IpElement extends KeyElement {
+
+        private final boolean self;
+        private final PlayerCommandElement possiblePlayer;
+
+        protected IpElement(Text key, boolean self) {
+            super(key);
+            this.self = self;
+            this.possiblePlayer = new PlayerCommandElement(key, false);
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            if (!args.hasNext() && this.self) {
+                if (source instanceof Player) {
+                    return ((Player) source).getConnection().getAddress().getAddress();
+                } else {
+                    throw args.createError(Text.of("No IP address was specified!"));
+                }
+            }
+            Object state = args.getState();
+            String s = args.next();
+            try {
+                return InetAddress.getByName(s);
+            } catch (UnknownHostException e) {
+                try {
+                    return ((Player) this.possiblePlayer.parseValue(source, args)).getConnection().getAddress().getAddress();
+                } catch (ArgumentParseException ex) {
+                    if (this.self && source instanceof RemoteSource) {
+                        args.setState(state);
+                        return ((RemoteSource) source).getConnection().getAddress().getAddress();
+                    }
+                    throw args.createError(Text.of("Invalid IP address!"));
+                }
+            }
+        }
+
+        @Override
+        public Text getUsage(CommandSource src) {
+            return src instanceof RemoteSource && this.self ? Text.of("[", super.getUsage(src), "]") : super.getUsage(src);
+        }
+
+    }
+
+    /**
+     * Expect an argument to return a {@link BigDecimal}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement bigDecimal(Text key) {
```

Comment:
Actually, it's `bigInteger` that is somewhat unnecessary - the point of `bigDecimal` is that the Economy API is entirely structured around them, and I find myself constantly using casts and checks (not to mention the fact that an arbitrary economy plugin may use enormous numbers).